### ISSUE TITLE
TypeError: unhashable type: 'Argh'

### ISSUE THREAD

**Issue by joernheissler** [python-trio/trio#1005]
```python3
#!/usr/bin/env python3.7

from typing import Any
import trio

class Argh(Exception):
    code: int

    def __init__(self, code: int) -> None:
        super().__init__()
        self.code = code

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Argh):
            return False

        return self.code == other.code

    def __repr__(self) -> str:
        return f'Argh(code={self.code})'

    def __str__(self) -> str:
        return repr(self)


async def die() -> None:
    raise Argh(844)


async def main() -> None:
    async with trio.open_nursery() as nursery:
        nursery.start_soon(die)


trio.run(main)
```

Results in:
```
Traceback (most recent call last):
  File "destroy.py", line 35, in <module>
    trio.run(main)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_run.py", line 1444, in run
    raise runner.main_task_outcome.error
  File "destroy.py", line 32, in main
    nursery.start_soon(die)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_run.py", line 496, in __aexit__
    combined_error_from_nursery = self._scope._close(new_exc)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_run.py", line 372, in _close
    exc = MultiError.filter(self._exc_filter, exc)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_multierror.py", line 209, in filter
    return _filter_impl(handler, root_exc)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_multierror.py", line 111, in _filter_impl
    push_tb_down(None, root_exc, preserved)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_multierror.py", line 104, in push_tb_down
    push_tb_down(new_tb, child_exc, preserved)
  File "/home/joern/.venv/trio/lib/python3.7/site-packages/trio/_core/_multierror.py", line 99, in push_tb_down
    if exc in preserved:
TypeError: unhashable type: 'Argh'
```

My Exception type is not hashable. In the real use case I get even uglier errors like `TrioInternalError`. Making the exception type hashable seems to resolve all issues.

But why does trio require hashable exceptions?

**Comment by smurfix** [comment]
> My Exception type is not hashable.
Why not? Simply add identity-based `__hash__` and `__eq__` functions.
> But why does trio require hashable exceptions?
Because `MultiError`. See the documentation in `trio/_core/_multierror.py` #16ff.

**Comment by njsmith** [comment]
Yeah, it's a bug/oversight in the MultiError code. Eventually #611 will replace all this, and @smurfix is right about the working, but it wouldn't hurt to fix it either. There's an incomplete patch at #767; maybe someone wants to pick it up and finish it?

**Comment by ziirish** [comment]
I might give it a look later this week if you want.

**Comment by njsmith** [comment]
Feel free!

**Comment by joernheissler** [comment]
> > My Exception type is not hashable.
> 
> Why not?

I used `@dataclass` for my exception class which makes it unhashable by default.
Now it's hashable, so problem fixed.

Real problem was that my program died a horrible death and at first nothing pointed at this unhashable exception type. Only after I added some debugging I was able to see this exception.

**Comment by oremanj** [comment]
Looks like this was resolved by #1023 -- please reopen if I'm missing something.
